### ISSUE TITLE
problem in first time setup

### ISSUE THREAD

**Issue by baskin** [webiny/webiny-js#776]
## This is:
- Bug

## Specifications
  - version: 3.2.5 (cli)
  - OS: mac
  - Browser: chrome

## Actual Behavior

tail of `webiny deploy-api --debug`
> webiny:formBuilder Finished API Gateway deployment: 'https://xxxxxxx.execute-api.us-east-1.amazonaws.com/prod' +1ms
webiny:gateway Inputs were not changed, no action required. +0ms
  webiny:api Input was not changed, no action required. +0ms
Error: invalid reference ${api.url}
    at Object.<anonymous> (/usr/local/lib/node_modules/@webiny/cli/sls/template/utils.js:33:27)
    at walker (/usr/local/lib/node_modules/@webiny/cli/node_modules/traverse/index.js:190:22)
    at /usr/local/lib/node_modules/@webiny/cli/node_modules/traverse/index.js:208:29
    at Array.forEach (<anonymous>)

## Steps to Reproduce the Problem

Team, attempted to setup webiny for the first time today.
1. Followed steps in https://docs.webiny.com/docs/get-started/quick-start/
2. All pre-reqs were done and tested
node -v: v12.16.1
yarn -v: 1.17.3
aws sts get-caller-identity: success
mongo connectivity verified
3. On running `webiny deploy-api --debug` get the above error.

Pl help.

**Comment by bigb123** [comment]
Hi @baskin,

As a temporary solution: can you try to use the current version of node.js (13.12.0 at the moment I'm writing it - https://nodejs.org/en/) rather than LTS? 

There was a problem reported on Webiny Community and the solution was to use the current version of the node.js - https://community.webiny.com/t/errors-trying-to-setup-local-webiny-environment/114/3. 

Also: I'm using yarn version 1.22.4

**Comment by baskin** [comment]
@bigb123 Tried both (upgraded node to 13.12.0 and yarn to 1.22.4).
Same issue `Error: invalid reference ${api.url}` still persists.

**Comment by Pavel910** [comment]
@baskin  I assume this log is from subsequent deploy attempts, not the first one. Because it says ‘no input was changed’. 

Try removing resources by running `webiny remove-api —debug` and then deploy again.

**Comment by baskin** [comment]
Thanks. I actually removed the entire setup and did it afresh with the above versions of node & yarn. Indeed the api deployment is now successful. Many thanks.

However, since then I tried moving to the next step of starting the admin app per https://docs.webiny.com/docs/get-started/quick-start/.
```
cd apps/admin
yarn start
```

The code compiles fine, but at runtime it errors (CORS) out saying
```
Access to fetch at 'https://<xxxx>.cloudfront.net/graphql' from origin 'http://localhost:3001' has been blocked by CORS policy:
```

I tried opening 'https://<xxxx>.cloudfront.net/graphql' separately and I get a 500 from cloudfront with a json response
```
{
  "requestContext": {
    "resourceId": "no2tvn",
    "resourcePath": "/graphql",
    "httpMethod": "GET",
    "extendedRequestId": "KJPlOGoNoAMFzTQ=",
    "requestTime": "29/Mar/2020:08:43:38 +0000",
    "path": "/prod/graphql",
    "accountId": "xxxxxx",
    "protocol": "HTTP/1.1",
    "stage": "prod",
    "domainPrefix": "xxxxxx",
    "requestTimeEpoch": 1585471418951,
    "requestId": "07a2fa0a-91eb-4ef1-a47e-c5fda5f78905",
    "domainName": "xxxxxx.execute-api.us-east-1.amazonaws.com",
    "apiId": "21utjgk5d0"
  },
  "context": {
    "callbackWaitsForEmptyEventLoop": true,
    "functionVersion": "$LATEST",
    "functionName": "74342c33_local_gateway_apollo-gateway",
    "memoryLimitInMB": "512",
    "logGroupName": "/aws/lambda/74342c33_local_gateway_apollo-gateway",
    "logStreamName": "2020/03/29/[$LATEST]9671e438d97f4597a5077f04d042ac8d",
    "invokedFunctionArn": "arn:aws:lambda:us-east-1:596207344607:function:xxxxxx_local_gateway_apollo-gateway",
    "awsRequestId": "92c26e5f-6980-4c06-ad97-bc6a717707e2"
  },
  "errors": [
    {
      "error": "Only absolute URLs are supported",
      "service": "security",
      "url": "undefined/graphql"
    }
  ]
}
```

**Comment by Pavel910** [comment]
The API deployment is definitely failing 🤔 When you open `https://xyz.cloudfront.net/graphql` it should open GraphQL Playground. The error you see is due to one of the services failing to deploy. Let me try creating a blank project; I'll get back to you.

**Comment by Pavel910** [comment]
@baskin so I just created a blank project and deployed it successfully. Could you post the content of this file: `api/.webiny/gateway/handler.js` ? We need to find out which service fails to deploy on your side.

**Comment by baskin** [comment]
Ok. Looks like "security" service failed in deployment somehow.
```
services: [
    {
      name: "security",
      url: "undefined/graphql"
    },
```

Rest looked fine
```
{
      name: "files",
      url: "https://28ct7bon33.execute-api.us-east-1.amazonaws.com/prod/graphql"
    },
    {
      name: "pageBuilder",
      url: "https://ppm2qmmqe2.execute-api.us-east-1.amazonaws.com/prod/graphql"
    },
    {
      name: "i18n",
      url: "https://9wpybhubkc.execute-api.us-east-1.amazonaws.com/prod/graphql"
    },
    {
      name: "formBuilder",
      url: "https://is6zsu3ov8.execute-api.us-east-1.amazonaws.com/prod/graphql"
    }
  ]
```

**Comment by Pavel910** [comment]
Ok, now try deploying just that one service using `webiny deploy-api --alias=gateway --alias=security --debug`, see if that goes well. So with this command you're deploying Apollo Gateway and Security services.

**Comment by baskin** [comment]
Should I do a remove-api first?

Running `webiny deploy-api --alias=gateway --alias=security --debug` as is, gave the same error.
Below is the relevant log.
` webiny:security Finished API Gateway deployment: undefined +2ms`
Below is the complete output
```
 webiny deploy-api --alias=gateway --alias=security --debug
💡 Loaded local environment from /Users/bhupi/repo/makeroo/webiny-cms/.env.json...
💡 Loaded local environment from /Users/bhupi/repo/makeroo/webiny-cms/api/.env.json...

  webiny:security Generating boilerplate code at '/Users/bhupi/repo/makeroo/webiny-cms/api/.webiny/security' +0ms
  webiny:security Start bundling with webpack +947ms
  webiny:security Finished bundling +7s
  webiny:security Deploy lambda +83ms
  webiny:security Starting deployment of lambda '74342c33_local_security_graphql' to the 'us-east-1' region. +104ms
  webiny:security No role provided for lambda '74342c33_local_security_graphql' +110ms
  webiny:security Syncing role 74342c33-ne5krd9 in region us-east-1. +8ms
  webiny:security Saved state for role 74342c33-ne5krd9. +5s
  webiny:security Role 74342c33-ne5krd9 was successfully deployed to region us-east-1. +0ms
  webiny:security Deployed role arn is arn:aws:iam::596207344607:role/74342c33-ne5krd9. +0ms
  webiny:security Packaging lambda code from '/Users/bhupi/repo/makeroo/webiny-cms/api/.webiny/security/build' +1ms
  webiny:security Successfully deployed lambda '74342c33_local_security_graphql' in the 'us-east-1' region. +3s
  webiny:security Deploying API Gateway +4ms
  webiny:security Input was not changed, no action required. +22ms
  webiny:security Finished API Gateway deployment: undefined +2ms
  webiny Loaded user ID: 136c867f-2e88-4797-a53d-aa5ef6502a73 +0ms
  webiny Tracking is DISABLED +0ms
  webiny:gateway Inputs were not changed, no action required. +0ms
  16s › gateway › Deploying ...
🎣 Processing stateChanged hook in admin app
🎣 Processing stateChanged hook in site app

🎉 Done! Deploy finished in 16s.
🚀 GraphQL API URL: https://d65aciexvqv49.cloudfront.net/graphql
```

**Comment by Pavel910** [comment]
This is very weird. I have no idea why it is failing, I'm also deploying all this back and forth as we speak and it's all working :( Could you try creating a completely new project?

**Comment by baskin** [comment]
Sure. Will try that and couple of other things and will report back in case it works. Thanks for the help thus far :)

**Comment by Pavel910** [comment]
Sorry you ran into all those problems :( Also if you want, join us on https://gitter.im/Webiny/webiny-js, it may be easier and faster to chat there.

**Comment by baskin** [comment]
I did a `webiny remove-api` and a redeploy worked like a charm. (except there was a minor hiccup of missing graphql endpoint on cloudfront -- figured that was due to delay in cloudfront propagation). 

I think a previous deployment failure would have rendered the config (e.g. url: "undefined/graphql") in an inconsistent state and a `remove-api` followed by a `deploy-api` fixed that. 

Thanks @Pavel910 & @bigb123 for your help on this. You can close this.

**Comment by Pavel910** [comment]
Glad you figured it out! 🚀 Let us know if you need help along the way :)

**Comment by bigb123** [comment]
Hi guys,

Just a quick comment on the end: I came across the `undefined` link problem before. It happened when I lost wifi connection for a moment during deployment. After reading debug log from deployment I realised some components had `undefined` links. My solution was to run `remove-api` and clean up `api/.serverless/` directory. 

My feeling is this error happens randomly when some errors occurred on the client side during deployment.

@Pavel910 thanks for a `--alias=` trick. It will speed up partial deployments.
